sno/checkout.py

```diff
@@ -178,27 +208,65 @@ def switch(ctx, create, force_create, discard_changes, refish):
         # Switch to existing branch
         #
         # refish could be:
-        # - branch name
+        # - local branch name eg 'master'
+        # - remote branch name eg 'master' (if do_guess is True)
+        #   (But not a remote branch eg 'origin/master')
         if not refish:
             raise click.UsageError("Missing argument: REFISH")
 
-        try:
-            branch = repo.branches[refish]
-        except KeyError:
+        if refish in repo.branches.remote:
+            # User specified something like "origin/master"
+            raise click.BadParameter(
+                f"A branch is expected, got remote branch {refish}",
+                param_hint="refish",
+            )
+
+        existing_branch = None
+        if refish in repo.branches.local:
+            existing_branch = repo.branches[refish]
+        elif do_guess:
+            # Guess: that the user wants create a new local branch to track a remote
+            existing_branch = _find_remote_by_name(repo, refish)
+
+        if not existing_branch:
             raise NotFound(f"Branch '{refish}' not found.", NO_BRANCH)
-        commit = branch.peel(pygit2.Commit)
 
+        commit = existing_branch.peel(pygit2.Commit)
         same_commit = repo.head.peel(pygit2.Commit) == commit
         if not discard_changes and not same_commit:
             ctx.obj.check_not_dirty(_DISCARD_CHANGES_HELP_MESSAGE)
 
+        if existing_branch.shorthand in repo.branches.local:
+            branch = existing_branch
+        else:
+            # Create new local branch to track remote
+            click.echo(
+                f"Creating new branch '{refish}' to track '{existing_branch.shorthand}'..."
+            )
+            branch = repo.create_branch(refish, commit)
+            branch.upstream = existing_branch
+
         head_ref = branch.name
 
     reset_wc_if_needed(repo, commit, discard_changes=discard_changes)
 
     repo.set_head(head_ref)
 
 
+def _find_remote_by_name(repo, name):
```

Comment:
I'd suggest calling this `_find_remote_branch_by_name`, since it finds a branch rather than a remote